syntax = "proto3";

package inventory;

service InventoryService {
  // Unary
  rpc GetItem(GetItemRequest) returns (GetItemResponse);
  rpc AddItem(AddItemRequest) returns (AddItemResponse);
  rpc UpdateItem(UpdateItemRequest) returns (UpdateItemResponse);
  rpc DeleteItem(DeleteItemRequest) returns (DeleteItemResponse);

  // Server Streaming
  rpc ListItems(ListItemsRequest) returns (stream ListItemResponse);

  // Client Streaming
  rpc BulkUpdateItems(stream BulkUpdateRequest) returns (BulkUpdateResponse);

  // Bi-Directional Streaming
  rpc MonitorInventory(stream MonitorInventoryRequest) returns (stream MonitorInventoryResponse);
}

// Common Data Structures
message Item {
  int32 item_id = 1;
  string name = 2;
  int32 quantity = 3;
}

// Unary Messages
message GetItemRequest {
  int32 item_id = 1;
}

message GetItemResponse {
  Item item = 1;
}

message AddItemRequest {
  string name = 1;
  int32 quantity = 2;
}

message AddItemResponse {
  Item item = 1;
}

message UpdateItemRequest {
  Item item = 1;
}

message UpdateItemResponse {
  Item item = 1;
}

message DeleteItemRequest {
  int32 item_id = 1;
}

message DeleteItemResponse {
  bool success = 1;
}

// Server Streaming Messages
message ListItemsRequest {}

message ListItemResponse {
  Item item = 1;
}

// Client Streaming Messages
message BulkUpdateRequest {
  int32 item_id = 1;
  int32 quantity_change = 2;
}

message BulkUpdateResponse {
  string message = 1;
}

// Bi-Directional Streaming Messages
message MonitorInventoryRequest {
  int32 item_id = 1;
}

message MonitorInventoryResponse {
  Item item = 1;
}